File: packages/ts/src/application/ports/llm-provider.ts
import type { Message } from "../../domain/conversation";
import type { ToolCall, ToolDefinition } from "../../domain/tool";

export interface CompletionRequest {
  readonly system: string;
  readonly messages: readonly Message[];
  readonly tools: readonly ToolDefinition[];
}

export interface CompletionResult {
  /** The model's text answer, or null when it only emitted tool calls. */
  readonly text: string | null;
  readonly toolCalls: readonly ToolCall[];
}

/**
 * Normalized, model-agnostic completion port (see ADR-0005). Each vendor is an
 * adapter that translates to and from this shape. One method, tool-use in and
 * tool-calls out.
 */
export interface LlmProvider {
  complete(request: CompletionRequest): Promise<CompletionResult>;
}

File: packages/py/src/airlock/application/ports/llm_provider.py
"""LlmProvider port: one normalized, model-agnostic completion call (ADR-0005)."""

from collections.abc import Sequence
from dataclasses import dataclass
from typing import Protocol

from airlock.domain.conversation import Message
from airlock.domain.tool import ToolCall, ToolDefinition


@dataclass(frozen=True)
class CompletionRequest:
    system: str
    messages: Sequence[Message]
    tools: Sequence[ToolDefinition]


@dataclass(frozen=True)
class CompletionResult:
    text: str | None
    tool_calls: tuple[ToolCall, ...]


class LlmProvider(Protocol):
    async def complete(self, request: CompletionRequest) -> CompletionResult: ...
